### ISSUE TITLE
[Dataset] Deal with large files

### ISSUE THREAD

**Issue by tanguyduval** [neuropoly/qMRLab#113]
I tried to upload a 200Mb NODDI dataset, but indeed that was blocked..
I propose that we use git lfs:
https://git-lfs.github.com/

does someone has any experience with this?

**Comment by tanguyduval** [comment]
done in branch Patch. Very handy and fast

**Comment by mathieuboudreau** [comment]
I've used it (git-lfs). Fair warning though, I believe the users will now have to download that very large file (and I believe, every copy of it if it gets changed in future commit), which means the *.zip files/clones might be a bit bloated. Just keep an eye on it to make sure we don't start making users download gigabyte+ transfers when they download qMRLab; we should keep the large files outside of the repo and in a separate location if they are not absolutely necessary and are only examples.

**Comment by mathieuboudreau** [comment]
Also, it's possible that  contributing to the git repo may have to have git-lfs installed on their computers (git-lfs is not git, it's an external extension). This issue mentions that: https://github.com/TeamPorcupine/ProjectPorcupine/issues/328

I already have git-lfs on my laptop so I can't test this hypothesis, maybe someone else in the group who doesn't have it installed could pull your branch and see if the file was downloaded.

If so, we will need to update the documentation to ensure people have git-lfs before cloning the repo, otherwise those large files might not be downloaded depending on their system's default git setups.

**Comment by tanguyduval** [comment]
Again, very relevant feedback Mathieu, thanks! :)
If git-lfs is not installed, files are replaced by a text file (containing a link) with the exact same name and extension.
We could create a function that checks if datasets are here. If not, download dataset using urlwrite. For instance `urlwrite https://github.com/neuropoly/qMRLab/blob/Patch/Data/IR_demo/IRData.nii.gz?raw=true test.nii.gz` download the Inversion Recovery data...
This function can be called in startup.m, or in each batch example..
Note that the data are automatically downloaded in releases, so I agree we should be careful to have small datasets.

**Comment by tanguyduval** [comment]
Apart from that git-lfs is easy to install and transparent afterwards. It prevents from having a growing repo.

**Comment by mathieuboudreau** [comment]
Ok thanks! The other issue is  related to Julien's email this morning. Github accounts have bandwidth limits. If the Github repo is 400mb in size, it only takes three fresh clones to go over the current neuropoly limit. I did a fresh clone of this repo last night, including your Patch branch (which contains duplicate copies of every single data file, since you changed their storage to git-lfs), which is likely what contributed to the neuropoly github account to go beyond its bandwidth limit.

**Comment by mathieuboudreau** [comment]
Also, if this repo becomes larger than the 1gb storage limit for free github accounts (note that every time a data file is change and saved in a commit, old files are not removed, they accumulate when using version control), people may not be able to fork this repo. It's another point to consider - I don't have admin privileges, so I can't see what Github considers to be the current size. However, when I pull all the branches, qMRLab takes up 1.03 GB of space on my laptop currently...

**Comment by mathieuboudreau** [comment]
As to your last point,

>Apart from that git-lfs ... prevents from having a growing repo.

This is not true, unfortunately. I skipped over this in my explanation above, but the repo will always increase in size. What git-lfs does is allow you to git pull large files, that will only pull the pointers to large files. If you have git-lfs installed, that then allows you to retrieve the files from those pointers. But all the data is still stored somewhere, and contributes to github's space limit/bandwidth.

**Comment by tanguyduval** [comment]
Ok, we'll have to move to OSF then... and clean the .nii.gz from the repo
In each batch_example, we need to add 
`urlwrite link filename`

**Comment by mathieuboudreau** [comment]
I'm not familiar with OSF, but I'll look into it, thanks!

To add to my last point, for reference, here is a link that explains how the limits are taken into account for git-lfs: https://help.github.com/articles/about-storage-and-bandwidth-usage/

Bandwith is not counted for git-lfs uploads, but are for downloads, which may be why you did not get a warning about the usage.

**Comment by mathieuboudreau** [comment]
Note that you may need to delete your branch and recreate it to properly remove the files, because simply deleting them and pushing the commit of that delete will keep the files from previous commits on the system/repo.

**Comment by tanguyduval** [comment]
data on https://osf.io/tmdfu/

**Comment by jcohenadad** [comment]
nice @tanguyduval, but don't you want to zip (or tar.gz) each folder, so that only a single download per model is required?

**Comment by tanguyduval** [comment]
Ok I deleted the branch and regenerated the commits in Patch2

**Comment by tanguyduval** [comment]
I agree, I will zip the files along with the fitresults

**Comment by jcohenadad** [comment]
anyone knows a way to download data from the OSF URL?  i'm looking for a matlab equivalent of curl/wget

**Comment by tanguyduval** [comment]
urlwrite
https://www.mathworks.com/help/matlab/ref/urlwrite.html

**Comment by tanguyduval** [comment]
@jcohenadad urlwrite is better than websave because not compatible by Octave.
Did you commited your modif? We are going to finish this with Agah

**Comment by tanguyduval** [comment]
@jcohenadad @agahkarakuzu 
I tried with octave, and `urlwrite('https://osf.io/zkefh/download?version=1','test.zip')` works fine.
On Matlab, use `websave('test.zip','https://osf.io/zkefh/download') `
To handle that we need to create a method `DownloadDataset` in the Abstract Class, and use `if Octave ...`

**Comment by tanguyduval** [comment]
See also my message in #139 
In the method downloadData (where you manage the condition octave OR Matlab) you can add a warning for old versions of Matlab

**Comment by tanguyduval** [comment]
Data are on OSF and can be downloaded smoothly locally.
However, there is a failure on TRAVIS--> no outputs from octave's urlwrite.
@jcohenadad are you aware of any difference between local and Travis for downloading data?

They had similar issue on SCT project, here is their script:  https://github.com/neuropoly/spinalcordtoolbox/blob/f03153603589723a790ccb66f7e40c8bac6fa1a2/scripts/sct_download_data.py

**Comment by jcohenadad** [comment]
sometimes the server is not responding. I would catch the exception and display it to understand what the problem is (runtime error? cannot connect to server? typo? permission error? etc.)

**Comment by tanguyduval** [comment]
I already tried in this commit 136fda2730394c052407741c05fbaa8188ee27b5
But no error message (SUCCESS = true). We need to find an alternative to urlwrite to debug this error

**Comment by jcohenadad** [comment]
strange... is anything downloaded?

**Comment by tanguyduval** [comment]
Absolutly no files are downloaded. 
Travis output:
````
Please wait. Downloading data...
before:
.   ..
after:
.   ..
````

**Comment by tanguyduval** [comment]
@agahkarakuzu maybe fork qMRLab to create a private repo --> this way you can debug Travis
https://docs.travis-ci.com/user/running-build-in-debug-mode/ 

Useful in the future..

**Comment by tanguyduval** [comment]
solved using curl 66bdc9b93bba67f35124515bf8dd1affaab1f58c
